ENH: Add `__class_getitem__` to `ndarray`, `dtype` and `number`

BvB93 opened the pull request:
This PR adds the `__class_getitem__` classmethod  to the `ndarray`, `dtype` and `number` classes, 
mimicking what [PEP 585](https://www.python.org/dev/peps/pep-0585/) did for the standard library. Consequently, expressions that were previously 
only allowed in .pyi stub files or with the help of `from __future__ import annotations` are now also 
legal during runtime.

A couple things of note (most of them summarized from the linked PEP):
* The new `__class_getitem__` method returns parametrized wrapper (`types.GenericAlias`) around the original type and, in most situations, can be used whenever the non-wrapped type is expected.
* The `types.GenericAlias` class is only available for for python >= 3.9, consequently `__class_getitem__` will only be available for these version.
* `types.GenericAlias` performs ~no~ minimal validation (_i.e._ checking the number of passed parameters) on the passed object, so this means that `__class_getitem__` can in practice take any object. I'm personally not a huge fan of this, but, on the other hand, I feel that copying the stdlib approach here is the lesser of evils due to its simplicity and maintainability, especially since the more thorough argument parsing performed by [`typing.Generic.__class_getitem__`](https://github.com/python/cpython/blob/7dacb70485a0910eb298c24b4d051720ca56fb91/Lib/typing.py#L1296-L1321) is quite complex and prone to change.
* Matching the [stub files](https://github.com/numpy/numpy/blob/dc7dafe70a53d6c122091516f34058bd0a6d89e1/numpy/__init__.pyi#L3252-L3255), the concrete `number` subclasses (`uint8`, `int32`, _etc._) are treated as if they were already subscribed, and will thus raise upon calling `__class_getitem__` due to a lack of free parameters.

Examples
----------
``` python
In [1]: import numpy as np
   ...: from typing import Any

In [2]: np.dtype[np.int64]
Out[2]: numpy.dtype[numpy.int64]

In [3]: np.inexact[Any]
Out[3]: numpy.inexact[typing.Any]

In [4]: np.ndarray[Any, np.dtype[np.float64]]
Out[4]: numpy.ndarray[typing.Any, numpy.dtype[numpy.float64]]

In [5]: np.float64[Any]
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-d62322c38e84> in <module>
----> 1 np.float64[Any]

TypeError: There are no type variables left in numpy.float64
```

Commit:
ENH: Add `ndarray.__class_getitem__`

BvB93 committed:
ENH: Add `number.__class_getitem__`

Commit:
ENH: Add `dtype.__class_getitem__`

Commit:
DOC: Add a release note for `__class_getitem__`

BvB93 commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -1805,6 +1805,19 @@ gentype_setflags(PyObject *NPY_UNUSED(self), PyObject *NPY_UNUSED(args),
     Py_RETURN_NONE;
 }
 
+/*
+ * Use for concrete np.number subclasses, making them act as if they
+ * were subtyped from e.g. np.signedinteger[object], thus lacking any
+ * free subscription parameters
+ */
+static PyObject *
+numbertype_class_getitem(PyObject *cls, PyObject *args)
+{
+    return PyErr_Format(PyExc_TypeError,
+                        "There are no type variables left in %s",
+                        ((PyTypeObject *)cls)->tp_name);
+}
+
```
Mimic the same error used by `types.GenericAlias`: https://github.com/python/cpython/blob/40d2ac92f9a28a486156dafdbb613016bb1f6b98/Objects/genericaliasobject.c#L302-L304

BvB93 commented on numpy/core/src/multiarray/descriptor.c:
```diff
@@ -3112,6 +3112,13 @@ static PyMethodDef arraydescr_methods[] = {
+
+    /* for typing; requires python >= 3.9 */
+    #ifdef Py_GENERICALIASOBJECT_H
+    {"__class_getitem__",
+        (PyCFunction)Py_GenericAlias,
+        METH_CLASS | METH_O, NULL},
```
For reference: the [Py_GenericAlias docs](https://docs.python.org/3/c-api/typehints.html).

seberg replied:
Probably, I am just lacking the typing knowledge, but why setting this to some generic (slightly weird?) error better than just leaving it undefined?

BvB93 replied:
My initial line of thought was that it would make `hasattr(x, "__class_getitem__")` checks more reliable, but after some further thinking I think a more explicit error message might be more desirable here. The conditionally generated documentation is not exactly helpful either (_e.g._ CircleCI runs on 3.8 it seems).

seberg replied:
Ah, the thing I utterly overlooked was that you want this on `np.generic` or at least `np.inexact` and then you have to override it.

BvB93 commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2756,6 +2756,13 @@ NPY_NO_EXPORT PyMethodDef array_methods[] = {
         (PyCFunction) array_format,
         METH_VARARGS, NULL},
 
+    /* for typing; requires python >= 3.9 */
+    #ifdef Py_GENERICALIASOBJECT_H
+    {"__class_getitem__",
+        (PyCFunction)Py_GenericAlias,
+        METH_CLASS | METH_O, NULL},
+    #endif
+
```
As `Py_GenericAlias` is only available for python >= 3.9, there are two ways of dealing with this limitation:
* Conditionally define the `__class_getitem__` method (as is currently done).
* Always define `__class_getitem__`, but let it raise for python 3.8.

I'm personally not a huge fan of methods that always raise, hence the first approach (this is open for discussion though).

BvB93 committed:
MAINT: Directly use `np.ndarray.__class_getitem__` for the definition of `npt.NDArray`

Commit:
MAINT: Make `__class_getitem__` available to all python version and perform basic validation of its input arguments

It will still raise on python 3.8, but now with a more explicit exception message

Commit:
TST,DOC: Skip the `__class_getitem__` doctests for python 3.8

BvB93 committed:
TST: Expand the `__class_getitem__` test coverage for python 3.8

BvB93 committed:
DOC: Add links to the `__class_getitem__` docs

BvB93 committed:
ENH: Add special-casing for `complexfloating` so that it can take 2 parameters

charris commented:
@BvB93 Do you consider this ready?

BvB93 commented:
Personally, I'm happy with the current state of the PR. I'm not sure if anyone else has some final remarks?

charris commented:
Let's give it a shot. Sounds like we will want it in Python >= 3.9 in any case.
